src/cudadev/CUDADataFormats/SiPixelClustersCUDA.h

```diff
@@ -4,12 +4,54 @@
 #include "CUDACore/device_unique_ptr.h"
 #include "CUDACore/host_unique_ptr.h"
 #include "CUDACore/cudaCompat.h"
+#include "DataFormats/SoALayout.h"
+#include "DataFormats/SoAView.h"
 
 #include <cuda_runtime.h>
 
 class SiPixelClustersCUDA {
 public:
-  SiPixelClustersCUDA() = default;
+  GENERATE_SOA_LAYOUT(DeviceLayoutTemplate,
+    SOA_COLUMN(uint32_t, moduleStart),  // index of the first pixel of each module
+    SOA_COLUMN(uint32_t, clusInModule), // number of clusters found in each module
+    SOA_COLUMN(uint32_t, moduleId),     // module id of each module
+
+    // originally from rechits
+    SOA_COLUMN(uint32_t, clusModuleStart) // index of the first cluster of each module
+  )
+  
+  // We use all defaults for the template parameters.
+  using DeviceLayout = DeviceLayoutTemplate<>;
+
+    GENERATE_SOA_VIEW(DeviceViewTemplate,
+    SOA_VIEW_LAYOUT_LIST(SOA_VIEW_LAYOUT(DeviceLayout, deviceLayout)),
+    SOA_VIEW_VALUE_LIST(
+      SOA_VIEW_VALUE(deviceLayout, moduleStart),  // index of the first pixel of each module
+      SOA_VIEW_VALUE(deviceLayout, clusInModule), // number of clusters found in each module
+      SOA_VIEW_VALUE(deviceLayout, moduleId),     // module id of each module
+  
+      // originally from rechits
+      SOA_VIEW_VALUE(deviceLayout, clusModuleStart) // index of the first cluster of each module
+    )
+  )
+  
+  using DeviceView = DeviceViewTemplate<>;
+  
+  GENERATE_SOA_CONST_VIEW(DeviceConstViewTemplate,
+    SOA_VIEW_LAYOUT_LIST(SOA_VIEW_LAYOUT(DeviceView, deviceView)),
+    SOA_VIEW_VALUE_LIST(
+      SOA_VIEW_VALUE(deviceView, moduleStart),  // index of the first pixel of each module
+      SOA_VIEW_VALUE(deviceView, clusInModule), // number of clusters found in each module
+      SOA_VIEW_VALUE(deviceView, moduleId),     // module id of each module
+  
+      // originally from rechits
+      SOA_VIEW_VALUE(deviceView, clusModuleStart) // index of the first cluster of each module
+    )
+  )
```

Comment:
Why is this "derived" from `DeviceView` and not from `DeviceLayout` ?

Reply:
The "idea" was that the class has its mutable view internally for initialization, and then down grades it to a const view for usage by the user in member function `view()`. Both path are possible, but we need to use the type we will pass to the constructor. The "hard work" is in the layout constructor, and in both cases, we're copying pointers (and the compiler will optimize this copying things out).

That makes the name of the `SOA_VIEW_LAYOUT_LIST` macro too restrictive, maybe. What about `SOA_VIEW_DATA_SOURCE_LIST` (and accordingly for `SOA_VIEW_LAYOUT`)?